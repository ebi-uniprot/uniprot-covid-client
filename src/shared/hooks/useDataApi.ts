import { useEffect, useReducer } from 'react';
import axios, { AxiosResponse, AxiosError } from 'axios';

import fetchData from '../utils/fetchData';

export type UseDataAPIState<T> = {
  loading: boolean;
  data?: T;
  status?: AxiosResponse['status'];
  statusText?: AxiosResponse['statusText'];
  headers?: AxiosResponse['headers'];
  error?: AxiosError;
  redirectedTo?: string;
};

enum ActionType {
  INIT = 'INIT',
  SUCCESS = 'SUCCESS',
  ERROR = 'ERROR',
}

type Action<T> =
  | { type: ActionType.INIT }
  | {
      type: ActionType.SUCCESS;
      response?: AxiosResponse<T>;
      originalURL?: string;
    }
  | { type: ActionType.ERROR; error: AxiosError };

// eslint-disable-next-line consistent-return
const createReducer = <T>() => (
  state: UseDataAPIState<T>,
  action: Action<T>
  // eslint-disable-next-line consistent-return
): UseDataAPIState<T> => {
  // eslint-disable-next-line default-case
  switch (action.type) {
    case ActionType.INIT:
      return {
        loading: true,
      };
    case ActionType.SUCCESS:
      // eslint-disable-next-line no-case-declarations
      const newState: UseDataAPIState<T> = {
        loading: false,
        data: action.response && action.response.data,
        status: action.response && action.response.status,
        statusText: action.response && action.response.statusText,
        headers: action.response && action.response.headers,
      };
      if (
        action.response &&
        action.response.request.responseURL &&
        action.response.request.responseURL !== action.originalURL
      ) {
        newState.redirectedTo = action.response.request.responseURL;
      } else if (
        // Issue with casing in axios-mock-adapter?
        action.response &&
        action.response.request.responseUrl &&
        action.response.request.responseUrl !== action.originalURL
      ) {
        newState.redirectedTo = action.response.request.responseUrl;
      }
      return newState;
    case ActionType.ERROR:
      return {
        loading: false,
        status: action.error.response && action.error.response.status,
        statusText: action.error.response && action.error.response.statusText,
        headers: action.error.response && action.error.response.headers,
        error: action.error,
      };
  }
};

function useDataApi<T>(url?: string | null): UseDataAPIState<T> {
  const [state, dispatch] = useReducer(createReducer<T>(), { loading: !!url });

  useEffect(() => {
    // need this variabe to ensure state updates don't occur when cancelled/unmounted
    // https://github.com/facebook/react/issues/14369#issuecomment-468267798
    let didCancel = false;

    // we don't require a URL, we just don't need data anymore
    // assume succes with no data
    if (!url) {
      dispatch({ type: ActionType.SUCCESS });
      return;
    }

    dispatch({ type: ActionType.INIT });

    // variables to handle cancellation
    const source = axios.CancelToken.source();

    // actual request
    fetchData<T>(url, undefined, source.token).then(
      // handle ok
      (response: AxiosResponse) => {
        if (didCancel) return;
        dispatch({ type: ActionType.SUCCESS, response, originalURL: url });
      },
      // catch error
      (error: AxiosError) => {
        if (axios.isCancel(error) || didCancel) return;
        dispatch({ type: ActionType.ERROR, error });
      }
    );

    // handle unmounting of the hook
    // eslint-disable-next-line consistent-return
    return () => {
      source.cancel();
      didCancel = true;
    };
  }, [url]);

  return state;
}

export default useDataApi;
