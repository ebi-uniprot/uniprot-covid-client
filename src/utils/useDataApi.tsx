/* eslint-disable @typescript-eslint/no-explicit-any */
import { useState, useEffect } from 'react';
import fetchData from './fetchData';

const useDataApi = (url: string) => {
  const [data, setData] = useState<any>({});

  useEffect(() => {
    const getData = async () => {
      try {
        const result = await fetchData(url);
        setData(result.data);
      } catch (error) {
        if (error.response.status !== '404') console.error(error);
        setData(null);
      }
    };
    getData();
  }, [url]);

  return data;
};

export default useDataApi;
