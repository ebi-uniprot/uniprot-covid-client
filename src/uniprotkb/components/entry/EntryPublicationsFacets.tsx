import React, { FC } from 'react';
import { Facets, Loader } from 'franklin-sites';
import { useLocation } from 'react-router-dom';

import { getUniProtPublicationsQueryUrl } from '../../config/apiUrls';

import useDataApiWithStale from '../../../shared/hooks/useDataApiWithStale';

import ErrorHandler from '../../../shared/components/error-pages/ErrorHandler';

import { getParamsFromURL } from '../../utils/resultsUtils';

import { Facet } from '../../types/responseTypes';

import './styles/entry-publications-facets.scss';

const EntryPublicationsFacets: FC<{ accession: string }> = ({ accession }) => {
  const { search } = useLocation();

  const { selectedFacets } = getParamsFromURL(search);
  const url = getUniProtPublicationsQueryUrl({
    accession,
    selectedFacets,
    size: 1, // TODO: change to 0 whenever the API accepts it
  });

  const { loading, data, status, error, isStale } = useDataApiWithStale<{
    facets: Facet[];
  }>(url);

  if (error) {
    return <ErrorHandler status={status} />;
  }

  if (loading && !data) {
    return <Loader />;
  }

  if (error || !data?.facets) {
    return <ErrorHandler status={status} />;
  }

  return (
    <div className={isStale ? 'is-stale' : undefined}>
      <Facets data={data.facets} />
    </div>
  );
};

export default EntryPublicationsFacets;
