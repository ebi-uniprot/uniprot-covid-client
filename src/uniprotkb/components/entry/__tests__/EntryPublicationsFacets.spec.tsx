import React from 'react';
import { fireEvent } from '@testing-library/react';
import EntryPublicationsFacets from '../EntryPublicationsFacets';
import renderWithRouter from '../../../../shared/__test-helpers__/RenderWithRouter';

jest.mock('../../../../shared/hooks/useDataApi', () => ({
  __esModule: true, // this makes it work
  default: jest.fn(() => ({
    loading: false,
    data: require('./__mocks__/entryPublicationsData.json'),
  })),
}));
import useDataApi from '../../../../shared/hooks/useDataApi';

describe('EntryPublication facets tests', () => {
  it('should render', () => {
    const { asFragment } = renderWithRouter(
      <EntryPublicationsFacets accession="P05067" />
    );
    expect(useDataApi).toHaveBeenCalled();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should add facet', () => {
    const { getByText, history } = renderWithRouter(
      <EntryPublicationsFacets accession="P05067" />
    );
    fireEvent.click(getByText(/Small/));
    const {
      location: { search },
    } = history;
    expect(search).toMatch(/facets=study_type:small_scale/);
  });

  it('should remove facet', () => {
    const { getByText, history } = renderWithRouter(
      <EntryPublicationsFacets accession="P05067" />,
      {
        route: '?facets=study_type:small_scale',
      }
    );
    fireEvent.click(getByText(/Small/));
    const {
      location: { search },
    } = history;
    expect(search).toMatch('');
  });
});
