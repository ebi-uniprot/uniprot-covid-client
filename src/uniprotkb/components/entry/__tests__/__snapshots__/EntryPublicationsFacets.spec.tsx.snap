// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryPublication facets tests should render 1`] = `
<DocumentFragment>
  <div>
    <div
      class="facets"
    >
      <ul
        class="no-bullet"
      >
        <li>
          <div
            class="facet-name"
          >
            Study type
          </div>
          <ul
            class="expandable-list no-bullet"
          >
            <li>
              <a
                href="/?facets=study_type%3Asmall_scale"
              >
                Small scale (239)
              </a>
            </li>
            <li>
              <a
                href="/?facets=study_type%3Alarge_scale"
              >
                Large scale (8)
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
